Better Multi-Language Support

UTDZac opened the pull request:
This PR offers a way for the Tracker to become fully translated in languages other than English. To accomplish this, anywhere and everywhere that text appears on the Tracker has been "extracted" and moved to a resource file for it's corresponding language.

The ultimate goal of this project is to get help from the community to translate the Tracker into different languages. This will be accomplished through a [GitHub Discussion forum](https://github.com/besteon/Ironmon-Tracker/discussions/389).

This PR is massive beyond reason and I don't expect a full review of everything. I would like some extra eyes on some of the larger-scale changes.

### Features
- 99.9% of all Tracker text that is displayed is now stored in their respective language file. Over 700 entries.
   - For example, All English resources can be found in `/Languages/English.lua`
- A new Language settings screen is available.
   - The display language of the Tracker can be changed independently of the language of the game rom being played (mostly, see known issues)
   - An option to "Autodetect language from game" is on by default, which will update the Tracker's display language based on the game being loaded.
- A new Release Notes page is available, which can be shown/hidden from the Update settings screen.
- _And maybe some other things I probably forgot. Been working on this one for a while._

### Primary Files to Focus Review
- `Resources.lua`
   - Change `DEBUG_REPLACE_ALL_RESOURCES` to `true` if you want a quick way to scan through the Tracker to make sure text resources are working and get replaced. Some names of things might not render or function properly with this enabled.
   - When you're done testing, let me know and I will remove the debug code, or you can do it.
- `Languages/English.lua`
- `data/RandomizerLog.lua` _(regex changes)_
- `screens/LanguageScreen.lua`
- `screens/UpdateScreen.lua`

### Questions
- Q1. Should I copy/paste all `English.lua` entries for all other Lua language files? I originally left them out, since there aren't any translations available, but perhaps it's easier to have all those text-keys already there instead of requiring others to add them in themselves. Granted, it will be *us* who eventually will add them all in, but if someone wants to translate on their own without our help, it might be easier if those entries were already there.
- **A1. I went with this solution. All entries from English.lua have been copied over to the other language lua files.**
- Q2. With regards to getting help from the community on translating the other languages, I'm unsure on the best way to present this info that will be translated. I was thinking having it split by priorities might be the best way to go? "Highest Priority" being the most common visible stuff, like the main tracker screen and settings, "Mid Priority" being other menus like Log Viewer, and "Lowest Priority" being rarely used texts such as move/ability info descriptions or form prompts.
- **A2. I'll be using the priority list system. Initially I'll just have the high-priority language entries that need translation online first. After some time, I'll add the other priority entries.**

### Notes & Known Issues
- Several MGBA resources are duplicates. This is intentional such that a translator can simply translate just the MGBA resources if they want to use those, and don't have to go hunting for the same resource hidden elsewhere.
- Some texts and resources are not translated. Usually this is because they are error messages that might appear before the `Resources` lua is loaded, or because I felt they were minor and not worth the time. Let me know if you think otherwise.
- **[FIXED]** If the display language of the Tracker doesn't match the game rom language, then the log viewer won't function properly, since it has to match the names of things in the log with the names of things in the Tracker.
- **[FIXED]** Bizhawk 2.8 encodes accented characters, and as such those won't properly be uppercase/lowercase in some places. This isn't an issue with Bizhawk 2.9 or MGBA.
- **[FIXED]** The LogViewer search screen currently doesn't work for accented characters. I couldn't find a clean solution for this, see `Utils.containsText()`
- While the Japanese language isn't official supported, the Tracker will allow loading a Japanese FireRed rom and playing it just fine.
- **[DEBUG REMOVED]** Please don't merge this PR in until the DEBUG stuff has been removed. It's only there to help with testing and reviewing for this PR.

Fellshadow reviewed (commented):
Seemed good from a skim through (though yeah too much to properly look through everything in a reasonable time). I did spot one thing relating to the animated pokemon popout, left a comment below

As for some thoughts on the questions / discussion points in the PR:

Q1: Yeah i think that would make sense to have placeholders there, though definitely would want to make clear that the text copy/pasted over is a placeholder and not that it happens to be the same between English / other language there

Q2: Not too sure myself either, given how much text I could see it being sorta messy regardless, but yeah I think splitting up into groups (be it priority groups or categories, perhaps both?) and  i think you could probably add each TODO item as tickboxes that we (or i guess whomever made the post with the TODO list) can check off as they are submitted / done, for people to easily see what is left?

> If the display language of the Tracker doesn't match the game rom language, then the log viewer won't function properly, since it has to match the names of things in the log with the names of things in the Tracker.

This one is tricky without some sort of big overhaul / rewrite of the log viewer. I think simplest solution may be to perhaps have like a `LogParserData` table in `Resources.lua` that contains the language data matching the loaded game needed for the log viewer to correctly parse the game's log file. Though I don't really know how much data this would be loading overall nor how much of an impact it could cause for people with perhaps lesser hardware or something.

At the very least, if not fixed then depending on how broken the log viewer is may need to throw an error or warning or something to tell the user "hey, this won't work right because X" but that's far from ideal

> The LogViewer search screen currently doesn't work for accented characters. I couldn't find a clean solution for this, see `Utils.startsOrContains()`

hm, maybe one approach could be a table mapping accented characters to a "closest matching non-accented character"? (e.g. é -> e and Á -> A etc) and using that to parse the text and replace accented characters with the mapped characters such that searching "Siccita" matches "Siccità" for example?

Fellshadow commented:
Also linking #62 as this is related and that issue is sorta our main "non-english support" update/progress log

UTDZac commented on ironmon_tracker/Drawing.lua:
```diff
@@ -511,12 +514,11 @@ function Drawing.setAnimatedPokemon(pokemonID)
 	if pokemonID ~= Drawing.AnimatedPokemon.pokemonID then
-		local pokemonData = PokemonData.Pokemon[pokemonID]
-		if pokemonData ~= nil then
+		local lowerPokemonName = Resources.Default.Game.PokemonNames[pokemonID]
```
These are defined in `English.lua` under GameResources > PokemonNames. Was this your question?

But yes, the "default" language is, and can only ever be, English.

Fellshadow replied:
I do see the `GameResources` section in `English.lua` however there isn't anything that loads it in to `Resources.Default.Game` or anything that defines `Resources.Default.Game`

From what I can see `Resources.Default` just contains `Language` and `Data` fields

UTDZac replied:
So I had to work a little Lua magic to get this setup. The goal was to be able to use `Resources.Game.PokemonNames` directly, without 1) having to worry about what language was loaded, the resources lookup should do that for you, and 2) without having to explicitly access `.Data`, which would always be required anyway.

The (1) is an fairly straightforward solution, such that anytime a new language is loaded I just replace the old stuff with the new stuff, and update resources across all globals that were using them. It's not perfect, but it works.

(2) I got around this annoyance by using meta tables. You can see this defined near the top of `Resources.lua`. This is basically just telling lua to use a shortcut: anytime `Resources.SOMETHING` is accessed/indexed, instead use `Resources.Data.SOMETHING`.

``` lua
-- Holds all current resources data used by all Tracker assets; data is categorized
Resources.Data = {}

-- Define metatables
local mt = {}
setmetatable(Resources, mt)
mt.__index = Resources.Data
```

Later on in `Resources.defineResourceCallbacks()` I declare that the "GameResources" table found in a language file, such as `English.lua`, is to be referenced as "Game" within the Resources lookup. Ultimately this results in a direct data insertion from `GameResources.PokemonNames` to `Resources.Game.PokemonNames`

With all that said, the `Resources.Default` data is merely the original resource data just copied over. When Resources initializes, it first loads English temporarily, copies over all of that data into `Resources.Default`, then later loads in the actual language data used by the selected language option (if different). Note: The `Resources.Default` has the same metatable magic as just regular `Resources`.

UTDZac commented:
Fixed the ability to search for names of things containing accented characters in https://github.com/besteon/Ironmon-Tracker/pull/390/commits/6f81cd6968a80c402a611ac4a95901930b8f3549 , using the keyboard that only has plain text characters. This required a good bit of reorganization work to get it done, partly why I avoided it in the first place, but had the energy today to get it resolved.

One lingering issue that was present before and after this change is that on Bizhawk 2.8, for names of things that start with special characters, they don't seem to render properly... and I'm not entirely sure why.

![image](https://github.com/besteon/Ironmon-Tracker/assets/4258818/2c4df7bb-a893-4605-8c46-4ccc1f50076c)

ÜBERWURF (Vital Throw) is not displayed correctly.

Fellshadow commented:
> Fixed the ability to search for names of things containing accented characters in [6f81cd6](https://github.com/besteon/Ironmon-Tracker/commit/6f81cd6968a80c402a611ac4a95901930b8f3549) , using the keyboard that only has plain text characters. This required a good bit of reorganization work to get it done, partly why I avoided it in the first place, but had the energy today to get it resolved.
> 
> One lingering issue that was present before and after this change is that on Bizhawk 2.8, for names of things that start with special characters, they don't seem to render properly... and I'm not entirely sure why.
> 
> ![image](https://user-images.githubusercontent.com/4258818/252824930-2c4df7bb-a893-4605-8c46-4ccc1f50076c.png)
> 
> ÜBERWURF (Vital Throw) is not displayed correctly.

@UTDZac 
Had a little dig on my bizhawk 2.8 install, seems it's coming from the `RandomizerLog.setupMappings` somehow failing to properly format the move name as lowercase (resulting in a failed move to id mapping later), which is weird because the same `toLowerUTF8` and `formatSpecialCharacters` Utils functions work fine when parsing the log? 

Attaching some Console output below, the umlaut characters print to the console weird but as you can see they get formatted fine in the end. Although this then makes me wonder if it's down to a difference in character encoding between our own files being read by `RandomizerLog.setupMappings` and the log file being read by e.g. `RandomizerLog.parseLog` and it just didn't affect the other characters because they happened to be lowercase to begin with
![image](https://github.com/besteon/Ironmon-Tracker/assets/106463662/bc2be294-621b-4f17-bf39-d7dd6c0e77f1)
![image](https://github.com/besteon/Ironmon-Tracker/assets/106463662/fa21d240-9740-4fad-869d-b492b32a2cd5)

UTDZac commented:
Fixed the Bizhawk 2.8 encoding issue in https://github.com/besteon/Ironmon-Tracker/pull/390/commits/4f892875c3e0ba3705e66c620f43726da22860bb by just adding duplicate entries to `Constants.Char` for each of the special characters that require encoding.

For example, instead of just one entry for `ý` there are now two:
```lua
	["ý"] = {
		encoded = "\253",
		upper = "Ý",
		plain = "y",
		width = 3,
	},
	["\253"] = {
		upper = "\221",
		plain = "y",
		width = 3,
	},
```

Fellshadow reviewed (approved):
Seems good to me, quickly checked on my end and the issue move from before loaded up fine this time
Hitting approve on this as a "i don't have any issues with this current state" but ofc there's still the debug stuff to take out whenever you feel it's ready to merge or if you want to add more stuff i can take a look again etc

UTDZac commented:
https://github.com/besteon/Ironmon-Tracker/pull/390/commits/2aee88c1b1a46029596afe9b7d1366a04a5fe888 Fixed the issue where the log couldnt be parsed if the Tracker's display language didnt match.

UTDZac commented:
- Fixed a few more issues relating to searching for names of things in the Log Viewer Search.
- The Log Viewer should now reflect language changes when the setting is changed. Works in most cases I could find.
- Added placeholder language resource entries for all non-English languages, noted by a trailing `-- NEEDS TRANSLATION`. It's not pretty, but it gets the job done.
- Removed old, redundant move and ability descriptions from their respective data files. Those descriptions now live in their respective language resource files
- Removed all references to `Constants.Words.POKEMON` and `.POKE`.

With all that said, I'm ready to merge this in. Just wanting to see if you have any final thoughts.

Fellshadow reviewed (approved):
Looks good to me, found a couple of typos in comments but that's it

UTDZac commented on ironmon_tracker/Languages/English.lua:
```diff
@@ -3042,7 +3036,7 @@ GameResources{
 	},
 	-- The list below must remain in the same order.
-	-- These are custom hand-written ability descriptions, only edit the "Desctipion" value
+	-- These are custom hand-written ability descriptions, only edit the "Desctipion" and "DescriptionEmerald" values
```
Fixed.

UTDZac commented on ironmon_tracker/screens/LogOverlay.lua:
```diff
@@ -1651,31 +1670,67 @@ function LogOverlay.buildTMGymButtons()
+	end
+end
+
+-- Rebuilds the buttons for the currently display screen. Useful when the Tracker's display language changes
```
Fixed.
